### ISSUE TITLE
Disable Python mode (torch dispatch mode) inside of mode-induced __torch_dispatch__call

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#75166]
### 🚀 The feature, motivation and pitch

When working with Python modes it is very easy to trigger infinite recursion, since most calls to any PyTorch function will end up back in the `__torch_dispatch__` for the mode. This makes, e.g., debug printing a pain, since you have to `no_dispatch` first.

I propose that we should switch the default around: by default, when we enter a `__torch_dispatch__` triggered by a mode, we should *disable* the mode (NB: this is different than `no_dispatch`!). If a user wishes `torch.add` to call back into `__torch_dispatch__`, they can explicitly reenable the mode (they can also explicitly recurse into themselves simply by invoking `self.__torch_dispatch__` directly). The argument is similar to the one posed at https://github.com/pytorch/pytorch/issues/75130 where we shouldn't reenable dispatcher features like autograd by default, and only reenable them if the user wants to reentrantly use the dispatcher on inner objects.

Note that there is a subtlety: I do not recommend we disable tensor-based torch dispatch (triggered by a tensor subclass which defines `__torch_dispatch__`), and the guidance here is irrelevant when considering a `__torch_dispatch__` induced by tensor-based dispatch (since there is no mode to disable). The reasoning here is twofold: (1) if a `__torch_dispatch__` mode is compositional, it may wish to trigger the `__torch_dispatch__` handlers of value objects after it is done performing its transformation--this would not happen by default if we disabled dispatch entirely with `no_dispatch`; (2) classic OO expectations demand that if you make calls on a subclass, those calls should dispatch to the subclass (and not the parent). Doing otherwise would violate this expectation. Instead, we should provide a way for people to, e.g., cast a subclassed tensor object back into an ordinary tensor, or balance at a knife's edge with `no_dispatch`.

This should probably be accompanied by a patch that makes torch dispatch mode work more like https://github.com/pytorch/pytorch/pull/75154 which has a more robust API that allows for chaining modes together.

cc @zou3519 @rgommers @peterbell10 @bdhirsh

**Comment by zou3519** [comment]
> I propose that we should switch the default around

To check my understanding, this would disable just the current mode, not all modes, right?

>  This makes, e.g., debug printing a pain, since you have to no_dispatch first.

Printing logic should probably do a no_dispatch at the very top (or there should be a way for _tensor_str to be overridden by Tensor subclasses), otherwise, things go through the default _tensor_str

**Comment by ezyang** [comment]
> To check my understanding, this would disable just the current mode, not all modes, right?

Yes; or more precisely, we would simply remove the mode from the mode stack. When there is only one mode (as in the current implementation), it would simply turn off the mode entirely.

**Comment by albanD** [comment]
I think this is a good idea.
The way I see it for modes, there are two ways to go:
- It is disabled by default when it is handled (this proposal) and the user have to explicitly re-enable it if they want it again.
- Nothing is done by default and the user has to disable this mode while handling to avoid infinite recursion (no_dispatch is NOT the same as this, this disables only the mode, we don't have an API to do this today). 

While the second one is closer to the way the "feature keys" in our dispatcher are implemented, the first one is definitely better to implement any feature that is not re-entrant (which is most of them and why we have so many TLS guards!).

**Comment by qqaatw** [comment]
I like this idea, do we have any update here?
